File: Backend/services/transactionservice/src/main/java/com/nexabank/transactionservice/entity/Transaction.java
package com.nexabank.transactionservice.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "transactions")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // The IBAN of the person sending money. It will be NULL if they are just doing a cash deposit at an ATM.
    @Column(nullable = true)
    private String senderIban;

    // The IBAN of the person receiving money. It will be NULL if they are doing an ATM withdrawal.
    @Column(nullable = true)
    private String receiverIban;

    @Column(nullable = false)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type; // DEPOSIT, WITHDRAWAL, TRANSFER

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionStatus status; // PENDING, COMPLETED, FAILED

    @Column(nullable = true)
    private String description; // Optional notes like "Rent Payment" or "ATM Withdrawal"

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }
}

File: frontend/types/transaction.ts
export type TransactionType = "DEPOSIT" | "WITHDRAWAL" | "TRANSFER";

export type TransactionStatus = "PENDING" | "COMPLETED" | "FAILED";

export type TransactionRequest = {
  amount: number;
  receiverIban?: string;
  description?: string;
};

export type TransactionResponse = {
  transactionId: number;
  type: TransactionType;
  status: TransactionStatus;
  amount: number;
  message: string;
  timestamp: string;
};

export type Transaction = {
  id: number;
  senderIban: string | null;
  receiverIban: string | null;
  amount: number;
  type: TransactionType;
  status: TransactionStatus;
  description: string | null;
  createdAt: string;
};
